src/topotoolbox/grid_object.py

```diff
@@ -302,6 +304,78 @@ def excesstopography(
 
         return result
 
+    def filter(self, method: str = 'mean', kernelsize: int = 3) -> 'GridObject':
+        """The function filter is a wrapper around various image filtering
+        algorithms. Only filters with rectangular kernels of uneven
+        dimensions are supported.
+
+        Parameters
+        ----------
+        method : str, optional
+            Which method will be used to filter the DEM: ['mean', 'average',
+            'median', 'sobel', 'scharr', 'wiener', 'std'], by default 'mean'
+        kernelsize : int, optional
+            Size of the kernel that will be applied. Note that ['sobel',
+            'scharr'] require that the kernelsize is 3, by default 3
+
+        Returns
+        -------
+        GridObject
+            The filtered DEM as a GridObject.
+
+        Raises
+        ------
+        ValueError
+            If the kernelsize does not match the requirements of this function
+            or the selected method is not implemented in the function.
+        """
+
+        valid_methods = ['mean', 'average', 'median',
+                         'sobel', 'scharr', 'wiener', 'std']
+
+        if method in ['mean', 'average']:
+            kernel = np.ones((kernelsize, kernelsize)) / kernelsize**2
+            filtered = convolve(self.z, kernel, mode='nearest')
+
+        elif method in ['median']:
+            filtered = median_filter(self.z, size=kernelsize, mode='reflect')
+
+        elif method in ['sobel', 'scharr']:
+            if kernelsize != 3:
+                arr = f"The method '{method}' only works with a 3x3 kernel'."
+                raise ValueError(arr) from None
+
+            if method == 'sobel':
+                ky = np.array([[1, 2, 1], [0, 0, 0], [-1, -2, -1]])
+            else:  # 'scharr'
+                ky = np.array([[3, 10, 3], [0, 0, 0], [-3, - 10, - 3]])
+
+            kx = ky.T
+            filtered = np.hypot(
+                convolve(self.z, ky, mode='nearest'),
+                convolve(self.z, kx, mode='nearest'))
+
+        elif method in ['wiener']:
+            filtered = wiener(self.z, mysize=kernelsize)
+
+        elif method in ['std']:
+            # This solution is based on this thread:
+            # https://stackoverflow.com/questions/19518827/what-is-the-python
+            # -equivalent-for-matlabs-stdfilt-function
+            filtered = generic_filter(self.z, np.std, size=kernelsize)
```

Comment:
MATLAB's `std` and numpy's `std` divide the sum of squares by different values to compute the sample variance and standard deviation (see [Bessel's correction](https://en.wikipedia.org/wiki/Bessel's_correction) for more details). To match the MATLAB behavior, you need to multiply the output by `np.sqrt(kernelsize**2 / (kernelsize**2 - 1))`.

Reply:
This is also incredibly slow. I think it is all right for now since there doesn't seem to be a better alternative readily available in `scipy`, but it is worth investigating in the future.